Code/Framework/AzCore/AzCore/Serialization/Json/MapSerializer.cpp

```diff
@@ -231,8 +231,29 @@ namespace AZ
             return context.Report(keyResult, "Failed to read key for associative container.");
         }
 
+        void* valueAddress = nullptr;
+        bool keyExists = false;
+        // Check if we should attempt to update the value if a key already exists
+        if (context.ShouldUpdateAssociativeContainer())
```

Comment:
What's the expected behavior for multi-map here? Seems like there could be some ambiguity. Should it just be disabled for multi-maps?

Reply:
Even the JSON standard is a bit undetermined on this. According to the standard https://datatracker.ietf.org/doc/html/rfc7159#section-4 "The names within an object SHOULD be unique.", but not that they "must" be unique. This means that implementations may have multiple keys with the same name, but how they're presented (the last only, the first only, an error, all values) is unpredictable. As such multi-maps are unpredictable. I think RapidJSON iterates over all keys even is there are duplicates and doesn't throw an error, but that's worth double checking. It's also a concern when moving away from RapidJSON in favor of the new Generic DOM (@NicholasVanSickle might also be interested in this discussion.) as that means that the behavior for this case can change based on what backend is used by the Generic DOM.
Multi-map in general isn't supported by JSON, so multi-map should probably be moved to an opt-in solution rather than be supported by default, but that can be a separate task as there will likely be some fallout from that. For this particular PR though I would stick with the same behavior as a regular map and update only the first element that's found. Right now that's also all that the Serialize Context will allow to be retrieved, so it's not even possible to get a value for a duplicate key.

Reply:
I think we should exclude support for multi-maps here and instead if the user needs to replicate support for storing a key that is mapped to multiple values, they would themselves use a container like `AZStd::(unordered_)map<KeyType>, AZStd::vector<ValueType>`.

We could even add a C++ specialization that `static_asserts` if an instantiation of `AZStd::unordered_multimap<K, V>` or `AZStd::map<K, V>` is attempted reflected at compile time.

Reply:
I think ScriptCanvas/GraphCanvas (de/)serializes unordered_multimap(like [here](https://github.com/o3de/o3de/blob/a10351f38ddf50af2b01ed071c04cda52d6069ed/Gems/GraphCanvas/Code/StaticLib/GraphCanvas/Types/GraphCanvasGraphSerialization.cpp#L37)). For the case of multimaps at least, I think we should keep the old behavior around(which is to add a duplicate element instead of updating the existing value). Adding @lsemp3d for ScriptCanvas. If we're going to make multimap support opt-in, maybe for the time being, the unordered map serializer expects some sort of meta-data to do the serialization? If it's not found, it asserts that unordered maps are an opt-in feature and does nothing.

Reply:
If we want to make multimap serialization opt-in and keep it around for legacy purposes only, do we need an RFC for that? I can write one up this week if needed. I was thinking of creating a `ContainerMetaData` class that would include `shouldClearContainer` and `shouldSupportMultimapSerialization` flags.

Reply:
As it stands now, multimap serialization is out-opt.
It doesn't really map well to JSON, it works in XML, because Nodes can have the same attributes and values.
If we mapped to a "key" -> "array of values" approach, then I think we can keep it as the default in general.
This is this the latter behavior that @AMZN-koppersr mentioned in his comment

Instead of ever supporting for multiple keys with the same value, we instead determine that if the container being loaded into is a (unordered_)multimap, then we check if the JSON value is an an array and the mapped type of the multimap can store an element of the array, than initialize the multimap using that JSON array.

Reply:
Sounds good. That's the current behavior so nothing more to do for this PR I guess.